File: php7/src/ChainOfResponsibility/Tests/Agent/AbstractLoanAgentTest.php
<?php

declare(strict_types = 1);

namespace PetrePatrasc\ChainOfResponsibility\Tests\Agent;
use PetrePatrasc\ChainOfResponsibility\Agent\AbstractLoanAgent;
use PetrePatrasc\ChainOfResponsibility\BankCustomer;
use PetrePatrasc\ChainOfResponsibility\LoanApplicationRequest;

/**
 * Governs all of the tests for the agents.
 *
 * @package PetrePatrasc\ChainOfResponsibility\Tests\Agent
 * @author Petre Pătraşc <petre@dreamlabs.ro>
 */
abstract class AbstractLoanAgentTest extends \PHPUnit_Framework_TestCase
{
    /**
     * Data provider for testing the business logic of each agent.
     *
     * @return array
     */
    abstract public function loanAmountDataProvider(): array;

    /**
     * Get the loan agent that is being tested.
     *
     * @return AbstractLoanAgent
     */
    abstract protected function getTestedAgent(): AbstractLoanAgent;

    /**
     * @param $loanAmount
     * @param $expectedApproval
     *
     * @dataProvider loanAmountDataProvider
     */
    public function testHandleLoanAmountByCurrentAgent(float $loanAmount, bool $expectedApproval)
    {
        $loanApplicationRequest = new LoanApplicationRequest(new BankCustomer("Unit Test Customer"), $loanAmount);
        $actualLoanResponse = $this->getTestedAgent()->handleLoanAmount($loanApplicationRequest);

        $this->assertNotNull($actualLoanResponse);
        $this->assertInstanceOf('\PetrePatrasc\ChainOfResponsibility\LoanApplicationResponse', $actualLoanResponse);

        $this->assertEquals($this->getTestedAgent(), $actualLoanResponse->getAgent());
        $this->assertEquals($expectedApproval, $actualLoanResponse->isApproved());
    }
}

File: java/src/ro/petrepatrasc/chainofresponsability/tests/agent/AbstractLoanAgentTest.java
package ro.petrepatrasc.chainofresponsability.tests.agent;

import com.tngtech.java.junit.dataprovider.DataProviderRunner;
import com.tngtech.java.junit.dataprovider.UseDataProvider;
import junit.framework.TestCase;
import org.junit.Test;
import org.junit.runner.RunWith;
import ro.petrepatrasc.chainofresponsability.agent.AbstractLoanAgent;
import ro.petrepatrasc.chainofresponsability.agent.TillAgent;
import ro.petrepatrasc.chainofresponsability.customer.BankCustomer;
import ro.petrepatrasc.chainofresponsability.loan.LoanApplicationRequest;
import ro.petrepatrasc.chainofresponsability.loan.LoanApplicationResponse;

@RunWith(DataProviderRunner.class)
abstract public class AbstractLoanAgentTest extends TestCase {

    @Test
    @UseDataProvider("loanAmountDataProvider")
    public void testHandleLoanAmountByCurrentAgent(double loanAmount, boolean expectedApproval) throws Exception {
        AbstractLoanAgent testedAgent = this.getTestedAgent();

        LoanApplicationRequest loanRequest = new LoanApplicationRequest(new BankCustomer(new String("Unit Test Customer")), loanAmount);

        LoanApplicationResponse actualLoanResponse = testedAgent.handleLoanRequest(loanRequest);

        this.assertEquals(testedAgent, actualLoanResponse.getAgent());
        this.assertEquals(expectedApproval, actualLoanResponse.isApproved());
    }

    /**
     * Get the loan agent that is being tested in the class.
     *
     * @return
     */
    abstract public AbstractLoanAgent getTestedAgent();
}
